src/compiler.ts

```diff
@@ -7718,9 +7919,109 @@ export class Compiler extends DiagnosticEmitter {
+      for (let _values = Map_values(instance.closedLocals), i = 0, k = _values.length; i < k; ++i) {
+        let local = unchecked(_values[i]);
+        members[i + 1] = this.program.makeNativeMember(local.name, local.type.intType.toString())
+      }
+
+      //Create a native class prototype with a dummy syntax tree, similar to native functions
+      var closureClassPrototype = assert(this.program.makeNativeClassPrototype(
+        "closure|" + this.program.nextClassId.toString(),
+        members
+      ));
+
+      //Resolve this prototype to get the class
+      var closureClass = this.resolver.resolveClass(closureClassPrototype, null)!;
+
+      //Compile this class to get the type
+      this.compileClass(closureClass);
+
+      //Append the appropriate signature and flags for this closure type, then set it to currentType
+      this.currentType = closureClass.type.asClosure(instance.signature)
+
+      //create a local which will hold our closure class instance
+      var tempLocal = flow.getTempLocal(this.currentType);
+
+      //copied closed locals into type
+      this.currentType.locals = instance.closedLocals;
+
+      var closureExpr = this.module.flatten([
+        this.module.local_set( //Allocate memory for the closure
+          tempLocal.index,
+          this.module.call(this.program.allocInstance.internalName, [
+            this.module.i32(closureClass.type.byteSize),
+            this.module.i32(closureClass.id)
+          ], NativeType.I32)
+        ),
+        this.module.store( //Store the function pointer at the first index
+          4,
+          this.module.local_get(tempLocal.index, this.options.nativeSizeType),
+          this.module.i32(index),
+          NativeType.I32,
+          0
+        ),
+        this.module.local_get(tempLocal.index, this.options.nativeSizeType) //load the closure locals index
+      ], this.options.nativeSizeType);
+
+      flow.freeTempLocal(tempLocal);
+
+      return closureExpr;
+    }
+
+    return this.module.i32(index);
+  }
+
+  private ifClosure(
+    indexExpr: ExpressionRef,
+    thenExpr: ExpressionRef,
+    elseExpr: ExpressionRef
+  ): ExpressionRef {
+    var module = this.module;
+
+    return module.if(
+      module.binary(
+        BinaryOp.EqI32,
+        module.binary(
+          BinaryOp.AndI32,
+          indexExpr,
+          module.i32(CLOSURE_FLAG)
+        ),
+        module.i32(CLOSURE_FLAG)
+      ),
+      thenExpr,
+      elseExpr
+    )
+  }
+
+  private getClosurePtr(closureExpr: ExpressionRef): ExpressionRef {
+    var module = this.module;
+
+    return module.binary(
+      BinaryOp.ShlI32,
+      closureExpr,
+      module.i32(4)
+    )
+  }
+
+  private toClosureReference(closureExpr: ExpressionRef): ExpressionRef {
+    var module = this.module;
+
+    return module.binary(
+      BinaryOp.OrI32,
+      module.binary(
+        BinaryOp.ShrI32,
+        closureExpr,
+        module.i32(4)
+      ),
+      module.i32(CLOSURE_FLAG)
+    )
```

Comment:
Helper methods for dealing with the new function pointer format, applying the appropriate shifts and setting the MSB etc

Reply:
Might belong elsewhere 🤷  maybe in `module.ts`? Though this isn't tied to binaryen very closely so maybe not..